UN-OCHA/unocha-site
# [UNO-616] Figures

Refs: UNO-616, UNO-618

This PR adds a `unocha_figures` module with notably a field type, widget and formatter for key figures.

The widget allows to select a provider from the OCHA Key Figures API and then drill down by country and year to select a figure or to select "manual" as provider and enter a label and value manually.

<img width="720" alt="Screenshot 2023-04-28 at 7 10 48" src="https://user-images.githubusercontent.com/696348/235011877-703f60ef-22d0-4bea-924b-ede8b02bf488.png">

*Note: the label is always overridable regardless of the provider.*

The formatter and related templates are relatively simple, showing the figures as a `<dl>` and formatting the number in a language aware manner either as decimal (thousand marker) or in compact form (1.2 million or 1.2M).

To select the figure format, 2 new view modes for the figures paragraph type have been introduced: Figures - Compact Long (1.2 million) and Figures - Compact Short (1.2M). The default view mode: "Figures" shows the numbers formatted using thousand markers (or whatever is appropriate to the selected language).

<img width="1054" alt="Screenshot 2023-04-28 at 7 09 32" src="https://user-images.githubusercontent.com/696348/235011903-5a4e45d9-e072-4783-a483-2518609bcc05.png">
<img width="1019" alt="Screenshot 2023-04-28 at 7 10 21" src="https://user-images.githubusercontent.com/696348/235011897-f233366d-012e-47e3-b95f-f9e9865db5b3.png">
<img width="980" alt="Screenshot 2023-04-28 at 7 09 59" src="https://user-images.githubusercontent.com/696348/235011901-63284bd3-e39e-4f66-835b-51037d030215.png">

There is also a drush command to update the figures from the OCHA Key Figures API that we can set up later to run at regular intervals.

## Deployment

To be deployed, this requires to first remove the existing "field_figures" (double field) from the figures paragraph type and then update the codebase.

## Notes

There is currently no data for OCT (OCHA funding, donors etc.). If that's not added to the OCHA Key Figures API, I'll add some logic for it in the unocha_figures module.

There is currently no automatic "Coverage" calculation. Once we know what kind of figures should be shown, we may be able to define more structured paragraph types and add extra processing to reduce the number of steps to pick up figures, show the coverage for funding figures etc.

We should remove the double field module at some point if we don't use it for something else.

This PR also contains fix for the caching of the various API clients as well as a fix for the buttons in a form dialog so that they stay were they are originally (ex: "add another item" at the bottom of the field it belongs to) instead of being moved to the "button" pane at the bottom of the dialog except for the main "save" and "cancel" button.

## Tests

1. Make a backup of your DB if you want to be able to revert.
2. Make sure to delete the `field_figures` on the Figures paragraph type before checking out this branch and importing the config.
3. Checkout branch, clear cache, import config (repeat if necessary)
4. Add the `$config['unocha_figures.settings']['ocha_key_figures_api_key'] = 'xxx';` to your local site settings.php, replacing `xxx` by the OCHA API key.
5. Create some node and add a Figures pragraph and some figures from various providers. Select your favorite view mode.
6. Check that the figures are correctly displayed.
